File: cpp/design_patterns/constructor/src/CarBasic.cpp
#include <stdio.h>
#include "CarBasic.h"

CarBasic::CarBasic(Engine* engine, Brakes* breaks):
        mEngine(engine),
        mBreaks(breaks)
{}

CarBasic::~CarBasic()
{
    delete mEngine;
    mEngine = NULL;

    delete mBreaks;
    mBreaks = NULL;
}

void CarBasic::Start() {
    printf("%s: Staring car\n", __func__);
    mEngine->Start();
    mBreaks->Release();
}

void CarBasic::Stop() {
    printf("%s: Stopping car\n", __func__);
    mBreaks->Apply();
    mEngine->ShutDown();
}

File: java/design/contructor/src/contructor/CarBasic.java
package contructor;

public class CarBasic implements Car {

  private Engine engine;
  private Breaks breaks;

  CarBasic(Engine engine, Breaks breaks) {
    this.engine = engine;
    this.breaks = breaks;
    
  }
  
  @Override
  public void Start() {
    System.out.println("Starting car");
    engine.Start();
    breaks.Release();
  }

  @Override
  public void Stop() {
    System.out.println("Stopping car");
    breaks.Apply();
    engine.Stop();
    
  }

}
